File: twentyone/1/one.py
def main():
    data = read_file("data.txt")

    print(sum(count_increases(data)))
    print(sum(count_increases_in_threes(data)))

def count_increases(data):
    prev = data[0]
    for i in range(1, len(data)):
        yield 1 if data[i] > prev else 0
        prev = data[i]

def count_increases_in_threes(data):
    prev_window = sum(data[:3])
    for to_add in range(3, len(data)):
        curr_window = prev_window - data[to_add - 3] + data[to_add]
        yield 1 if curr_window > prev_window else 0
        prev_window = curr_window

def read_file(filename):
    with open(filename, 'r') as f:
        lines = f.readlines()
    
    return [int(item.strip()) for item in lines]


if __name__ == "__main__":
    main()

File: twentyone/1/one.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func readData(filename string) []int64 {
	data, _ := ioutil.ReadFile(filename)

	result := make([]int64, 0)
	for _, v := range strings.Split(string(data), "\n") {
		if v == "" {
			continue
		}
		val, _ := strconv.ParseInt(v, 10, 0)
		result = append(result, val)
	}

	return result
}

func countIncreases(readings []int64) int {
	increases := 0
	for idx := 1; idx < len(readings); idx++ {
		if readings[idx] > readings[idx-1] {
			increases = increases + 1
		}
	}

	return increases
}

func countIncreasesInThree(readings []int64) int {
	increases := 0
	prevWindow := readings[0] + readings[1] + readings[2]
	for i := 3; i < len(readings); i++ {
		currentWindow := prevWindow - readings[i-3] + readings[i]
		if currentWindow > prevWindow {
			increases++
		}
	}
	return increases

}

func main() {
	readings := readData("data.txt")

	fmt.Println(countIncreases(readings))
	fmt.Println(countIncreasesInThree(readings))

}
